lib/pbench/agent/tool_meister.py

```diff
@@ -1296,7 +1290,7 @@ def stop_tools(self, data: dict) -> int:
 
         # Clean up the running tools data structure explicitly ahead of
         # potentially receiving another start tools.
-        for name in sorted(self._tools.keys()):
+        for name in sorted(self._usable_tools.keys()):
             try:
                 del self._running_tools[name]
             except KeyError:
```

Comment:
There shouldn't be any entries in `_running_tools` which aren't in `_usable_tools`, right?  And, we probably don't want any entries left in `_running_tools` after we're done with this loop, right?

So why not replace this loop with something like
```
        self._running_tools = {}
```
(If you cannot make that change, please consider beefing up the comment.)

Reply:
I'll noodle about this a bit more.

Reply:
Still noodling?

Reply:
Yes.

Reply:
Actually, I suspect this is one of the few cases where an `assert` might really make sense. As Webb points out, since "unusable tools" shouldn't be running, iterating through the (potentially) smaller set of "usable tools" makes sense, but we also expect that "running tools" will become empty by the end.

Instead of Webb's suggestion (which you didn't take) of arbitrarily setting `_running_tools` to empty, you might end the loop here by *asserting* that it's become empty... which always ought to be true, represents an internal tool class consistency error if not true, and isn't an error in the caller that needs to be reported as there's nothing to be done about it. (That is, the theoretically fatal assert error is OK for debugging and evaporating the check for production deployment -- if we ever do that -- would make sense.)

Reply:
The "usable" tools are the sub-set of originally requested tools which pass the `.install()` calls (see https://github.com/distributed-system-analysis/pbench/blob/d8c16a3d4206a7135404eb467540c189304c0367/lib/pbench/agent/tool_meister.py#L905).

By definition, the "running" tools are the sub-set of transient tools which start successfully (see https://github.com/distributed-system-analysis/pbench/blob/d8c16a3d4206a7135404eb467540c189304c0367/lib/pbench/agent/tool_meister.py#L1208).

We could assert that the set of running tools is a sub-set of the transient tools and then just remove it.

Reply:
Let's continue any further refactoring in PR #2725.